packages/@aws-cdk/cdk/lib/aspects/tag-aspect.ts

```diff
@@ -0,0 +1,149 @@
+import { Resource, TagType } from '../cloudformation/resource';
+import { Construct, IAspect } from '../core/construct';
+
+export interface TagAspectProps extends TagProperties {
+  // TODO docs
+  value?: string;
+}
+
+export interface TagProperties {
+  // TODO docs
+  propagate?: number;
+  include?: string[];
+  exclude?: string[];
+}
+
+export abstract class TagBaseAspect implements IAspect {
+
+  public readonly type: string = 'taggable';
+
+  public readonly key: string;
+  public readonly value?: string;
+
+  public propagate: number;
+  private readonly include: string[];
+  private readonly exclude: string[];
+  private readonly visitedBy: {[id: string]: boolean} = {};
+
+  private parent?: Construct;
+
+  constructor(key: string, props: TagAspectProps = {}) {
+    this.key = key;
+
+    this.propagate = props.propagate === undefined ? -1 : props.propagate;
+    this.value = props.value;
+    this.include = props.include || [];
+    this.exclude = props.exclude || [];
+  }
+
+  public visit(node: Construct): void {
+    if (this.visitedBy[node.uniqueId] === true) {
+      return;
+    }
+    this.visitedBy[node.uniqueId] = true;
+    const propagate = this.propagate > 0 ? this.propagate - 1 : this.propagate;
+    this.parent = this.parent || node;
+    if (propagate !== 0) {
+      for (const child of node.children) {
+        this.propagate = propagate;
+        child.apply(this);
```

Comment:
Agreed, I think the framework should recursively apply.

Reply:
Also, in my head the Tags where physically attached only to the construct the user `apply()`d them to, and only during `visit()`ation would they also `visit()` downward in the tree. Because... (:asterisk:, see below)

Reply:
I agree with all of these comments, but some of the comments conflict.  If `visit()` recurses then I think I need some form of a collector on the resources because mutating `properties` we've also agreed is not acceptable. That brings back something like `TagManager`. If I recurse in `apply` at the `Construct` level then our simplified ordering story *️⃣ below I think becomes difficult again. Is there something I'm missing that avoids these draw backs?

I leaning towards `visit` moving the tag down the tree and combining `TagType` with a simplified `TagManager` ideally named just `Tags`.